File: solutions/javascript/square-root/1/square-root.js
/**
 * Calculates the square root of a non-negative number using the Newton-Raphson method.
 * Truncates the result to a given number of decimal digits (no rounding).
 *
 * @param {number} x - The number to compute the square root of.
 * @param {number} [tolerance=1e-10] - Precision tolerance for convergence.
 * @param {number} [maxIterations=1000] - Maximum number of iterations to avoid infinite loops.
 * @returns {number} The truncated square root of x.
 * @throws {Error} If x is negative.
 */
export function squareRoot(x, tolerance = 1e-10, maxIterations = 1000) {
    if (x < 0) {
        throw new Error("Negative input is not supported.");
    }

    if (x < 2) {
        return x;
    }

    let guess = x / 2;
    let iterations = 0;

    while (Math.abs(guess * guess - x) > tolerance && iterations < maxIterations) {
        guess = (guess + x / guess) / 2;
        iterations++;
    }

    return guess | 0;
};

File: solutions/python/square-root/3/square_root.py
def square_root(number, tolerance=1e-8):
    """Compute the square root of a given non-negative number using Newton's method.

    This function estimates the square root by iteratively improving the guess
    until it converges to within a specified tolerance. The algorithm uses
    an initial guess equal to the input number and refines this guess
    through the formula: better_guess = 0.5 * (guess + number / guess).

    Args:
        number (float): The number to compute the square root of. Must be non-negative.
        tolerance (float): The tolerance level for determining convergence. Default is 1e-8.

    Returns:
        float: The estimated square root of the number.

    Raises:
        ValueError: If the input number is negative, indicating that the square root
        is not defined for negative values.
    """
    if number == 0:
        return 0
    if number < 0:
        raise ValueError("Cannot compute the square root of a negative number.")

    guess = number

    while True:
        better_guess = 0.5 * (guess + number / guess)

        if abs(guess - better_guess) < tolerance:
            return better_guess

        guess = better_guess
